bin/src/commands.rs

```diff
@@ -20,3 +26,41 @@ pub async fn ui_error(message: String, _stack: Option<Vec<String>>) -> AppResult
 
     Ok(())
 }
+
+#[tauri::command]
+pub async fn add_contract(
+    chain_id: u64,
+    address: Address,
+    db: tauri::State<'_, Db>,
+) -> AppResult<()> {
+    let provider = provider(chain_id as u32).await.unwrap();
+    let proxy = ethui_proxy_detect::detect_proxy(address, &provider)
+        .await
+        .unwrap()
+        .map(|_| address);
+
+    let (abi, name) = if networks_is_dev().await? {
+        (None, None)
+    } else {
+        let name = fetch_etherscan_contract_name(chain_id.into(), address).await?;
+        let abi = fetch_etherscan_abi(chain_id.into(), address)
+            .await?
+            .map(|abi| serde_json::to_string(&abi).unwrap());
+
+        (abi, name)
+    };
+
+    db.insert_contract_with_abi(chain_id as u32, address, abi, name, proxy)
+        .await?;
+
+    let maybe_proxy = ethui_proxy_detect::detect_proxy(proxy.unwrap(), &provider)
+        .await
+        .unwrap();
+
+    if maybe_proxy.is_some() {
+        let _ = Box::pin(add_contract(chain_id, proxy.unwrap(), db)).await;
+    }
```

Comment:
you're calling detect_proxy on the proxy again, but I'm not sure why

it's confusing because as I said above, this address is currently the proxy itself when it should be the impl. but overall I don't see why you'd need to call `detect_proxy` twice before recursing